patroni/postgresql/__init__.py

```diff
@@ -889,38 +889,44 @@ def remove_data_directory(self):
             logger.exception('Could not remove data directory %s', self._data_dir)
             self.move_data_directory()
 
-    def pick_synchronous_standby(self, cluster):
+    def _get_synchronous_commit_param(self):
+        query = "SELECT setting from pg_settings where name = 'synchronous_commit'"
+        return self.query(query).fetchone()[0]
+
+    def pick_synchronous_standby(self, cluster, sync_node_count=1):
         """Finds the best candidate to be the synchronous standby.
 
         Current synchronous standby is always preferred, unless it has disconnected or does not want to be a
         synchronous standby any longer.
 
         :returns tuple of candidate name or None, and bool showing if the member is the active synchronous standby.
         """
-        current = cluster.sync.sync_standby
-        current = current.lower() if current else current
         members = {m.name.lower(): m for m in cluster.members}
         candidates = []
-        # Pick candidates based on who has flushed WAL farthest.
-        # TODO: for synchronous_commit = remote_write we actually want to order on write_location
+        sync_nodes = []
+        # Pick candidates based on who has higher replay/remote_write/flush lsn.
+        sync_commit_par = self._get_synchronous_commit_param()
+        sort_lsn_col = {'remote_apply': 'replay', 'remote_write': 'write'}.get(sync_commit_par, 'flush')
+        sort_lsn_col = '{0}_{1}'.format(sort_lsn_col, self.lsn_name)
         for app_name, state, sync_state in self.query(
                 "SELECT pg_catalog.lower(application_name), state, sync_state"
                 " FROM pg_catalog.pg_stat_replication"
-                " ORDER BY flush_{0} DESC".format(self.lsn_name)):
+                " WHERE state = 'streaming'"
+                " ORDER BY sync_state DESC, {0} DESC".format(sort_lsn_col)):
```

Comment:
Sure.  The sort option ``ORDER BY sync_state DESC`` is with the intent to keep the Query result predicable.  Especially in cases when a synchronous standby member is slowed down OR async node receiving changes faster than the sync member (very rare but possible). Such cases would trigger sync standby members swapping frequently and we want to avoid that by keeping the query result predictable.

Reply:
I mean it deserves a comment in the code, because we are relying on very specific naming of sync_state: https://github.com/zalando/patroni/pull/1594#discussion_r443129801

Reply:
fixed